### PR TITLE
Pagination for QueryTableChanges

### PR DESCRIPTION
**Pull request by charlenelyu-db**
- [Part 1](https://github.com/delta-io/delta-sharing/pull/352): Pagination for QueryTable at snapshot
  - [Part 2](https://github.com/delta-io/delta-sharing/pull/353): Pagination for QueryTable from starting_version
    - [Part 3](https://github.com/delta-io/delta-sharing/pull/354): Pagination for QueryTableChanges
      - [Part 4](https://github.com/delta-io/delta-sharing/pull/356): Client change [[files changed](https://github.com/delta-io/delta-sharing/pull/356/files/c7271be600f1e8645d53a7c12d2578cceda4deee..b05024e06af268ecb7c2f9bd541071be3eb319ee)]

---

Support pagination for QueryTableChanges in reference server.

This is part 3 for issue https://github.com/delta-io/delta-sharing/issues/351.

### CODE REVIEW

File: server/src/main/scala/io/delta/standalone/internal/DeltaSharedTableLoader.scala
Comment (chakankardb) [inline server/src/main/scala/io/delta/standalone/internal/DeltaSharedTableLoader.scala:611]: Question: It seems first we will generate all the changes, and then apply pagination? So pagination will not save any processing on the server side right? Is the same true for DS server in UC as well? @linzhou-db

File: server/src/main/scala/io/delta/standalone/internal/DeltaSharedTableLoader.scala
Comment (charlenelyu-db) [reply server/src/main/scala/io/delta/standalone/internal/DeltaSharedTableLoader.scala:659]: Yeah I'm stacking PRs, sorry for the confusion. Please use the [files changed] link in the description to view incremental changes. I will clean things up before merging.

File: server/src/main/scala/io/delta/standalone/internal/DeltaSharedTableLoader.scala
Comment (charlenelyu-db) [reply server/src/main/scala/io/delta/standalone/internal/DeltaSharedTableLoader.scala:733]: By override I meant we should use (start, end) from the page token instead of the original request. I will rephrase it.

File: server/src/main/scala/io/delta/standalone/internal/DeltaSharedTableLoader.scala
Comment (charlenelyu-db) [reply server/src/main/scala/io/delta/standalone/internal/DeltaSharedTableLoader.scala:611]: We're saving some process: the `start` will be equal to the `startingVersion` in the page token, so we won't unpack delta logs for versions that are already processed in previous pages.

However, we still generate all changes until the `endingVersion` each time. I guess this part can be optimized in UC. Two options I can think of: 
1. we could push down `actionListener` to `getChanges()`, so we read logs and process actions together and it'll return in the middle
2. we don't change the interface of `getChanges()`, we call it to unpack one version at a time and process version by version.

WDYT?

Comment (chakankardb) [reply server/src/main/scala/io/delta/standalone/internal/DeltaSharedTableLoader.scala]: I think both options should work; the first one is similar to what we do for normal query, so may be preferable for consistency reason.
In the future, we will likely process json files in parallel as well (similar to checkpoint parquet).

Let's open a ticket for now to address this in UC.

**Review by chakankardb — APPROVED** [review]
Perhaps wait for Lin's review as well.

**Review by linzhou-db — APPROVED** [review]

File: server/src/main/scala/io/delta/standalone/internal/DeltaSharedTableLoader.scala
Comment (linzhou-db) [reply server/src/main/scala/io/delta/standalone/internal/DeltaSharedTableLoader.scala]: let's have a separate discussion about how to handle this in UC.

**Pull request merged**
